### DISCUSSION TITLE
Proposal to reduce memory usage of metric aggregation

### DISCUSSION THREAD

**Discussion by alangy98** [open-telemetry/opentelemetry-cpp#3061] [Ideas]
Hello,

We are currently working on reducing the redundant storage of `MetricAttribute`. In the current implementation, the entire `MetricAttributes` is stored in an `unordered_map`. Labels with lower cardinality are stored multiple times alongside labels with higher cardinality. Additionally, attributes cannot be shared across different metrics, even though using the same or similar `MetricAttributeSet` to update multiple metrics is common in real-world scenarios.

Our preliminary work is to adopt the idea of `HPACK`, using a global table to minimize redundant storage of label keys and values. Would the community be receptive to these changes?

**Comment by lalitb** [comment]
@alangy98 Yes indeed, any such change to reduce memory usage would be welcome, as long as any such global structure won't bring further mutex locks in the hot path of instrumentation.

ps. There are also few places of heap allocations which can be optimized, but that could be different topic :)
